subprojects/dependency-management/src/main/java/org/gradle/internal/component/model/VariantWithOverloadAttributes.java

```diff
@@ -21,15 +21,21 @@
 public class VariantWithOverloadAttributes implements VariantResolveMetadata.Identifier {
     private final VariantResolveMetadata.Identifier variantIdentifier;
     private final ImmutableAttributes targetVariant;
+    private final int hashCode;
 
     public VariantWithOverloadAttributes(VariantResolveMetadata.Identifier variantIdentifier, ImmutableAttributes targetVariant) {
         this.variantIdentifier = variantIdentifier;
         this.targetVariant = targetVariant;
+        this.hashCode = computeHashCode();
```

Comment:
This class is used specifically as a key in a HashMap. This is a common optimization used for this situation since the hash code is guaranteed to be called, likely several times. Though, it does come at the cost of increased memory usage. 

In practice, this pattern should really only be used along with profiling to ensure it actually improves performance. I added this earlier when attempting to avoid a performance regression. I can revert this and see if performance is still acceptable

Reply:
Leaving this as-is. The memory impacts are negligible and the hash code is guaranteed to be called.